crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -217,139 +231,90 @@ where
         &self,
         piece_getter: &PG,
         worker_state: &mut CacheWorkerState,
-        new_piece_caches: Vec<Arc<dyn PieceCache>>,
+        mut new_piece_caches: Vec<Arc<dyn PieceCache>>,
     ) where
         PG: PieceGetter,
     {
         info!("Initializing piece cache");
-        // Pull old cache state since it will be replaced with a new one and reuse its allocations
-        let cache_state = mem::take(&mut *self.piece_caches.write().await);
-        let mut stored_pieces = Vec::with_capacity(new_piece_caches.len());
-        let mut free_offsets = Vec::with_capacity(new_piece_caches.len());
-        for mut state in cache_state {
-            state.stored_pieces.clear();
-            stored_pieces.push(state.stored_pieces);
-            state.free_offsets.clear();
-            free_offsets.push(state.free_offsets);
+        const MAX_CACHES_NUM: usize = u16::MAX as usize;
```

Comment:
I'd create a type alias `type CacheIndex = u16` and used it both here (you'll be able to move this constant out BTW) and in `FarmerCacheOffset` data structure, such that reviewer can see that those things are in fact related more easily. For farms we even have a generic that is `u8` for local farms and `u16` for cluster setup because type size here impacts RAM usage.